### ISSUE TITLE
Feature: padding

### ISSUE THREAD

**Issue by qqpann** [qqhann/KnowledgeTracing#22]
## 概要
paddingを行うには，次の変更を行わなければいけない．
- dataloaderを作成する際にsequence_sizeに満たないものをbatchに含めるようにする．
- モデルがbatchに含まれたpadded sequenceを正しく処理できるようにする．
    - lossで正しく処理できるようにする．
    - aucで正しく処理できるようにする．

### ✘ pad_sequence

pack_padded_sequenceとか pad_sequence を使えという記述が多い．
しかし，その返り値はDataSetじゃないので書き換えるのが大変そう
- https://towardsdatascience.com/taming-lstms-variable-sized-mini-batches-and-why-pytorch-is-good-for-your-health-61d35642972e
maskをとるだけでうまくいくよという記述もある．手軽そうだが，lossの計算がそれでどううまくいってるのかちょっと謎い．

### ✘ ignore_index parameter for loss function
lossの初期化の時に，ignore_indexでpadを指定できるという記述をいくつか見かけた．
しかしこの方法はlossによって対応していないものがある．loss function依存のもの．（[NLLLoss](https://pytorch.org/docs/stable/nn.html?highlight=nllloss#torch.nn.NLLLoss)にはあって[BCELoss](https://pytorch.org/docs/stable/nn.html#bceloss)にはない）
また，そもそもこの使い方は，`"<PAD>"`の辞書インデックスを指定するという意味合いだと思われ，今回の実装とは合わない．
- <https://discuss.pytorch.org/t/calculating-loss-on-sequences-with-variable-lengths/9891/2> NLLLossを使えと．
- <https://discuss.pytorch.org/t/ignore-index-in-the-cross-entropy-loss/25006> ignore_iundexはBCEにはない．
- <https://discuss.pytorch.org/t/dataloader-for-various-length-of-data/6418> Dataloaderが可変長を扱うには?→pad_sequenceなど．　今回は型式が特殊だったため，結局listで手動でpadした．

- <https://discuss.pytorch.org/t/simple-working-example-how-to-use-packing-for-variable-length-sequence-inputs-for-rnn/2120/17> 可変長をRNNなどで扱う例についてのdiscussion.
- <https://github.com/HarshTrivedi/packing-unpacking-pytorch-minimal-tutorial> pad_packedなどの使い方が良くまとまっている．

### ✔ mask
sequenceのpad以外の部分にtrueフラグを立てて，使う時にフラグが立った場所だけにmaskするという方法．
- <https://towardsdatascience.com/taming-lstms-variable-sized-mini-batches-and-why-pytorch-is-good-for-your-health-61d35642972e> pytorchのfunctionによらず，手動でmaskを行った記事．この記事がどうしてうまくいくのかはよくわかりにくい．
- <https://pytorch.org/docs/stable/torch.html?highlight=masked_select#torch.masked_select> masked_select function

## 方法
masked_selectを使うことにする．
これを行うには：
- dataloaderを生成する際にyseq分のmaskを生成する．
- lossを計算する際にmasked_selectを行ったpredictionとtargetを与える．
- aucを計算する際にmasked_selectを行ったpredictionとtargetを与える．

必要がある．

注意点は，masked_selectの返り値はflattenなベクトルになること．
